calico_mesos/calico_mesos.py

```diff
@@ -399,11 +400,11 @@ def _reserve(hostname, uid, ipv4_addrs, ipv6_addrs):
         for ip_addr in ipv4_addrs + ipv6_addrs:
             datastore.assign_ip(ip_addr, uid, {}, hostname)
             # Keep track of succesfully assigned ip_addrs in case we need to rollback
-    except (RuntimeError, ValueError):
+    except (RuntimeError, ValueError, AlreadyAssignedError):
         failed_addr = ip_addr
         _log.error("Couldn't reserve %s. Attempting rollback." % (ip_addr))
         # Rollback assigned ip_addrs
-        datastore.release_ips(ipv4_addrs + ipv6_addrs)
+        datastore.release_ips(set(ipv4_addrs + ipv6_addrs))
```

Comment:
I think you'll need to handle this case separately: If the IP is already assigned we don't want to release it; something else is using it!

You'll need to keep track of what IPs we successfully assign, so that we can back out just those assignments in the case of failure.